datalake/src/main/java/com/sequenceiq/datalake/flow/datalake/upgrade/preparation/DatalakeUpgradePreparationActions.java

```diff
@@ -0,0 +1,151 @@
+    public Action<?, ?> finishedAction() {
+        return new AbstractSdxAction<>(SdxEvent.class) {
+
+            @Inject
+            private SdxStatusService sdxStatusService;
+
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext, SdxEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, SdxEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Sdx upgrade was finalized with sdx id: {}", payload.getResourceId());
+                sdxStatusService.setStatusForDatalakeAndNotify(
+                        DatalakeStatusEnum.RUNNING,
+                        ResourceEvent.DATALAKE_UPGRADE_PREPARATION_FINISHED,
+                        "Upgrade preparation finished",
+                        payload.getResourceId());
+                sendEvent(context, DATALAKE_UPGRADE_PREPARATION_FINALIZED_EVENT.event(), payload);
+            }
+
+            @Override
+            protected Object getFailurePayload(SdxEvent payload, Optional<SdxContext> flowContext, Exception ex) {
+                LOGGER.error("Datalake upgrade preparation finalization failed: {}", payload, ex);
+                return DatalakeUpgradeFailedEvent.from(payload, ex);
+            }
+        };
+    }
+
+    @Bean(name = "DATALAKE_UPGRADE_PREPARATION_FAILED_STATE")
+    public Action<?, ?> failedAction() {
+        return new AbstractSdxAction<>(DatalakeUpgradePreparationFailedEvent.class) {
+
+            @Inject
+            private SdxStatusService sdxStatusService;
+
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    DatalakeUpgradePreparationFailedEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, DatalakeUpgradePreparationFailedEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Sdx upgrade preparation failed for sdxId: {}", payload.getResourceId());
+                sdxStatusService.setStatusForDatalakeAndNotify(
+                        DatalakeStatusEnum.DATALAKE_UPGRADE_PREPARATION_FAILED,
+                        "Upgrade preparation failed: " + payload.getException().getMessage(),
+                        payload.getResourceId());
+                sendEvent(context, DATALAKE_UPGRADE_PREPARATION_FAILED_HANDLED_EVENT.event(), payload);
+            }
+
+            @Override
+            protected Object getFailurePayload(DatalakeUpgradePreparationFailedEvent payload, Optional<SdxContext> flowContext, Exception ex) {
+                LOGGER.error("Upgrade preparation flow failure handling failed unexpectedly!", ex);
+                return null;
```

Comment:
is this ok to be null?

Reply:
yes, failed state should not throw any exception, we usually have this in these states